### PR TITLE
rac2,kvserver: expose RACv2 status in ranges.json

### PR DESCRIPTION
**Pull request by pav-kv**
This PR plumbs the basic information about the RACv2 range controller and send streams to the `RangeInfo` proto exposed via `ranges.json` in `debug.zip`.

Part of #136014

### CODE REVIEW

File: pkg/server/serverpb/status.proto [hunk 1]
```diff
@@ -382,6 +382,18 @@ message RaftState {
   util.hlc.Timestamp lead_support_until = 8 [(gogoproto.nullable) = false];
 }
 
+// RACState contains the state of the Replication Admission Control component of
+// a range.
+message RACState {
```
Comment (pav-kv) [inline pkg/server/serverpb/status.proto:387]: What should we include here? I'm thinking both the `RangeController` state (such as `nextRaftIndex`, `forceFlushIndex`), and per-stream (tokens held, `indexToSend`).

File: pkg/server/status.go [hunk 2]
```diff
@@ -2520,6 +2523,7 @@ func (s *systemStatusServer) rangesHelper(
 		return serverpb.RangeInfo{
 			Span:          span,
 			RaftState:     convertRaftStatus(raftStatus),
+			RacState:      rac2State,
```
Comment (pav-kv) [inline pkg/server/status.go:2526]: nit: add a proto directive to make this named as `RACState`

**Comment by pav-kv** [comment]
Example:

```json
    "rac_state": {
      "streams": {
        "1": {
          "index_to_send": 1708,
          "next_raft_index": 1708,
          "tokens_held": [
            0,
            131,
            0,
            0
          ]
        },
        "2": {
          "index_to_send": 1708,
          "next_raft_index": 1708,
          "tokens_held": [
            0,
            131,
            0,
            0
          ]
        },
        "3": {
          "index_to_send": 1708,
          "next_raft_index": 1708,
          "tokens_held": [
            0,
            131,
            0,
            0
          ]
        }
      }
    },
```

**Comment by pav-kv** [comment]
@kvoli @sumeerbhola This is a draft, please take an initial look.

- Does the plumbing make sense, or there is a more canonical way? Maybe we could integrate with `InspectRaftMuLocked` better?
- What do we want to include in this json?

**Review by sumeerbhola — COMMENTED** [review]
looks good

> Reviewed 10 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-:-OEP_EtEEu_EMz9ohqAN:b-47fafm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-OEPXQlj1-HsHhYEFAQ5:-OEPXQlj1-HsHhYEFAQ6:bh994ju) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada0ef65c443e201a65fe1a4545d315c1dec4923/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L124)):*
> ```Go
> 	//
> 	// Requires replica.raftMu to be held.
> 	StatusRaftMuLocked() serverpb.RACState
> ```

nit: the method is called "Status" and the proto says "State". How about picking one term for both.
___
*[`pkg/server/serverpb/status.proto` line 391 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-OEPZNi718qkbMcnQgZO:-OEPZNi718qkbMcnQgZP:bg8m2rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada0ef65c443e201a65fe1a4545d315c1dec4923/pkg/server/serverpb/status.proto#L391)):*
> ```proto
>     uint64 index_to_send = 1;
>     uint64 next_raft_index = 2;
>     repeated int64 tokens_held = 3;
> ```

send_tokens_held

File: pkg/server/status.go (hunk 2 above)
Comment (pav-kv) [reply pkg/server/status.go]: Done

File: pkg/server/serverpb/status.proto (hunk 1 above)
Comment (pav-kv) [reply pkg/server/serverpb/status.proto]: Done

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137680)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-OEPXQlj1-HsHhYEFAQ5:-OEU9E3u6VFz0SGf0rGM:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada0ef65c443e201a65fe1a4545d315c1dec4923/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L124)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: the method is called "Status" and the proto says "State". How about picking one term for both.
</blockquote></details>

Done
___
*[`pkg/server/serverpb/status.proto` line 391 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-OEPZNi718qkbMcnQgZO:-OEU9Ip13CT7CYCIPluT:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada0ef65c443e201a65fe1a4545d315c1dec4923/pkg/server/serverpb/status.proto#L391)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

send_tokens_held
</blockquote></details>

Done

**Comment by pav-kv** [comment]
Example after latest changes:

```json
    "rac_status": {
      "next_raft_index": 19829,
      "force_flush_index": 17,
      "streams": {
        "1": {
          "index_to_send": 19829,
          "next_raft_index_initial": 19193,
          "next_raft_index": 19829,
          "eval_tokens_held": [
            264,
            0
          ],
          "send_tokens_held": [
            0,
            264,
            0,
            0
          ]
        },
        "2": {
          "index_to_send": 19829,
          "next_raft_index_initial": 19195,
          "next_raft_index": 19829,
          "eval_tokens_held": [
            132,
            0
          ],
          "send_tokens_held": [
            0,
            132,
            0,
            0
          ]
        },
        "3": {
          "index_to_send": 19829,
          "next_raft_index_initial": 19195,
          "next_raft_index": 19829,
          "eval_tokens_held": [
            132,
            0
          ],
          "send_tokens_held": [
            0,
            132,
            0,
            0
          ]
        }
      }
    },
```

**Review by sumeerbhola — APPROVED** [review]
I am not experienced with `ranges.json` -- is this change adding a significant amount of state to what was there in ranges.json? For instance, if we are doubling the volume of ranges.json, we could try to prune something from here.

:lgtm: 

> Reviewed 1 of 10 files at r3, 8 of 9 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-:-OEUFtF3-z7vt6Gk4pyD:b-pwtole)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons, @kvoli, and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 1693 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-OEUFM91EFIP2-6nht8f:-OEUFM91EFIP2-6nht8g:b-9wtv5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/422929b13a403a3a35f2061b4b9d95e9fb5aab33/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L1693)):*
> ```Go
> func convertTokensSlice(tokens []kvflowcontrol.Tokens) []int64 {
> 	result := make([]int64, len(tokens))
> 	for pri, tok := range tokens {
> ```

nit: the index is not always a priority, since it can also be a WorkClass, so perhaps `i` instead of `pri`.

**Review by kvoli — APPROVED** [review]
The size increase is approx 14%, when looking at a ranges.json, single range entry and adding @pav-kv's json above:

```
1.1K Dec 19 15:18 new_range.json
7.6K Dec 19 15:18 current_range.json

8.7/7.6=1.14 14% inc
```

Seems fine to me and well worthwhile, but 14% slower would be noticeable. We should rely less on these eventually so :lgtm: 

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-:-OEVXveuAfX7VA29V-Pa:b-4ndwxd)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @angles-n-daemons, @pav-kv, and @sumeerbhola)

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137680)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @angles-n-daemons and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 1693 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/137680#-OEUFM91EFIP2-6nht8f:-OE_gxML46KF-pznAWAY:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/422929b13a403a3a35f2061b4b9d95e9fb5aab33/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L1693)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: the index is not always a priority, since it can also be a WorkClass, so perhaps `i` instead of `pri`.
</blockquote></details>

done

**Comment by pav-kv** [comment]
TFTRs! Merging for now, and we [can sort out](https://cockroachlabs.slack.com/archives/C06UFBJ743F/p1734725983930519) the size question in a follow-up.

bors r=sumeerbhola,kvoli

**Pull request merged**